lib/srv/db/redis/engine.go

```diff
@@ -0,0 +1,468 @@
+// a Redis instance. It's also responsible for audit.
+func (e *Engine) process(ctx context.Context, redisClient redis.UniversalClient) error {
+	for {
+		// Read commands from client.
+		cmd, err := e.readClientCmd(ctx)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+
+		e.Audit.OnQuery(e.Context, e.sessionCtx, common.Query{Query: cmd.String()})
+
+		e.Log.Debugf("redis cmd: %s", cmd.String())
+
+		// send valid commands to Redis instance/cluster.
+		err = e.processCmd(ctx, redisClient, cmd)
+
+		var vals interface{}
+		if _, ok := err.(redis.Error); ok {
+			vals = err
+		} else if errors.Is(err, context.DeadlineExceeded) {
+			// Do not return Deadline Exceeded to the client as it's not very self-explanatory.
+			// Return "connection timeout" as this is what most likely happened.
+			vals = errors.New("connection timeout")
+		} else if err != nil {
+			// Send the error to the client as connection will be terminated after return.
+			e.SendError(err)
+
+			return trace.Wrap(err)
+		} else {
+			vals, err = cmd.Result()
+			if err != nil {
+				return trace.Wrap(err)
+			}
+		}
+
+		// Send response back to the client.
+		if err := e.sendToClient(vals); err != nil {
+			return trace.Wrap(err)
+		}
+	}
+}
+
+// writeCmd writes Redis commands passed as vals to Redis wire form.
+// Most types is covered by go-redis implemented WriteArg() function. Types override by this function are:
+// * Redis errors and Go error: go-redis returns a "human-readable" string instead of RESP compatible error message
+// * integers: go-redis converts them to string, which is not always what we want to.
+// * slices: arrays are recursively converted to RESP responses.
+func writeCmd(wr *redis.Writer, vals interface{}) error {
+	switch val := vals.(type) {
+	case redis.Error:
+		if val == redis.Nil {
+			// go-redis returns nil value as errors, but Redis Wire protocol decodes them differently.
+			// Note: RESP3 has different sequence for nil.
+			if _, err := wr.WriteString("$-1\r\n"); err != nil {
+				return trace.Wrap(err)
+			}
+			return nil
+		}
+
+		if err := wr.WriteByte('-'); err != nil {
+			// Redis error passed from DB itself. Just add '-' to mark as error.
+			return trace.Wrap(err)
+		}
+
+		if _, err := wr.WriteString(val.Error()); err != nil {
+			return trace.Wrap(err)
+		}
+
+		if _, err := wr.Write([]byte("\r\n")); err != nil {
+			return trace.Wrap(err)
+		}
+	case error:
+		if _, err := wr.WriteString("-ERR "); err != nil {
+			// Add error header specified in https://redis.io/topics/protocol#resp-errors
+			// to follow the convention.
+			return trace.Wrap(err)
+		}
+
+		if _, err := wr.WriteString(val.Error()); err != nil {
+			return trace.Wrap(err)
+		}
+
+		if _, err := wr.Write([]byte("\r\n")); err != nil {
+			return trace.Wrap(err)
+		}
+	case int:
+		if err := writeInteger(wr, int64(val)); err != nil {
+			return trace.Wrap(err)
+		}
+	case int8:
```

Comment:
Does the driver not already provide this functionality which we could potentially expose/reuse and not reimplement in Teleport?

Reply:
Technically `redis.Writer.WriteArg()` provides such functionality, but it doesn't implement all types. I explain that in a comment at the top of this function.